File: topics/mpi/practicals/05.MPI_Topo/fortran/2.exchange_ghost_cells_solution.f90
PROGRAM ghost_cell_exchange

! Use only 16 processes for this exercise
! Send the ghost cell in two directions: left<->right and top<->bottom
! ranks are connected in a cyclic manner, for instance, rank 0 and 12 are connected
! process decomposition on 4*4 grid 
!  |-----------|
!  | 0| 4| 8|12|
!  |-----------|
!  | 1| 5| 9|13|
!  |-----------|
!  | 2| 6|10|14|
!  |-----------|
!  | 3| 7|11|15|
!  |-----------|

! Each process works on a 10*10 (SUBDOMAIN) block of data
! the d corresponds to data, g corresponds to "ghost cells"
! and x are empty (not exchanged for now)

!  xggggggggggx
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  xggggggggggx

  USE MPI
  IMPLICIT NONE

  INTEGER SUBDOMAIN, DOMAINSIZE
  PARAMETER (SUBDOMAIN = 10)
  PARAMETER (DOMAINSIZE = SUBDOMAIN+2)
  INTEGER rank, size, i, j, rank_right, rank_left, ierror
  DOUBLE PRECISION data(DOMAINSIZE,DOMAINSIZE)
  INTEGER request
  INTEGER status(MPI_STATUS_SIZE)
  INTEGER rank_top, rank_bottom
  INTEGER dims(2), periods(2)
  INTEGER comm_cart
  INTEGER data_ghost

  CALL MPI_Init(ierror)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierror)
  CALL MPI_Comm_size(MPI_COMM_WORLD, size, ierror)

  IF (size.NE.16) THEN
     WRITE (*,*)"please run this with 16 processors"
     CALL MPI_Finalize(ierror);
     STOP
  END IF
  DO i=1, DOMAINSIZE, 1
    DO j=1, DOMAINSIZE, 1
      data(i,j)=rank
    END DO
  END DO

  ! neighbouring ranks with cartesian grid communicator
  ! we do not allow the reordering of ranks here
  dims(1)=4
  dims(2)=4
  periods(1)=1
  periods(2)=1

  ! an alternative solution would be to allow the reordering and to use the new communicator for the communication
  ! then the MPI library has the opportunity to choose the best rank order with respect to performance
  ! CREATE a cartesian communicator (4*4) with periodic boundaries and use it to find your neighboring
  ! ranks in all dimensions.
  CALL MPI_Cart_create(MPI_COMM_WORLD, 2, dims, periods, 0, comm_cart, ierror)
  CALL MPI_Cart_shift(comm_cart, 0, 1, rank_left, rank_right, ierror)
  CALL MPI_Cart_shift(comm_cart, 1, 1, rank_top, rank_bottom, ierror)

  ! derived datatype, create a datatype to send the rows
  CALL MPI_Type_vector(SUBDOMAIN, 1, DOMAINSIZE, MPI_DOUBLE, data_ghost, ierror)
  CALL MPI_Type_commit(data_ghost, ierror)

  !  ghost cell exchange with the neighbouring cells in all directions
  !  to the left
  CALL MPI_Irecv(data(2,DOMAINSIZE), SUBDOMAIN, MPI_DOUBLE, rank_right, 0, MPI_COMM_WORLD, request, ierror)
  CALL MPI_Send(data(2,2), SUBDOMAIN, MPI_DOUBLE, rank_left, 0, MPI_COMM_WORLD, ierror)
  CALL MPI_Wait(request, status, ierror)

  !  to the right
  CALL MPI_Irecv(data(2,1), SUBDOMAIN, MPI_DOUBLE, rank_left, 0, MPI_COMM_WORLD, request, ierror)
  CALL MPI_Send(data(2,DOMAINSIZE-1), SUBDOMAIN, MPI_DOUBLE, rank_right, 0, MPI_COMM_WORLD, ierror)
  CALL MPI_Wait(request, status, ierror)

  !  to the top
  CALL MPI_Irecv(data(DOMAINSIZE,2), 1, data_ghost, rank_bottom, 0, MPI_COMM_WORLD, request, ierror)
  CALL MPI_Send(data(2,2), 1, data_ghost, rank_top, 0, MPI_COMM_WORLD, ierror)
  CALL MPI_Wait(request, status, ierror)

  !  to the bottom
  CALL MPI_Irecv(data(1,2), 1, data_ghost, rank_top, 0, MPI_COMM_WORLD, request, ierror)
  CALL MPI_Send(data(SUBDOMAIN+1,2), 1, data_ghost, rank_bottom, 0, MPI_COMM_WORLD, ierror)
  CALL MPI_Wait(request, status, ierror)

  IF (rank.EQ.9) THEN
     WRITE (*,*) 'data of rank 9 after communication'
     DO i=1, DOMAINSIZE, 1
        DO j=1, DOMAINSIZE, 1
          WRITE (*,'(F6.1)',advance='no') data(i,j)
        END DO
        WRITE (*,*)
     END DO
  END IF

  CALL MPI_Type_free(data_ghost, ierror)
  CALL MPI_Comm_free(comm_cart, ierror)
  CALL MPI_Finalize(ierror)

END PROGRAM

File: topics/mpi/practicals/05.MPI_Topo/cxx/2.exchange_ghost_cells_solution.c
/****************************************************************
 *                                                              *
 * This file has been written as a sample solution to an        *
 * exercise in a course given at the CSCS Summer School.        *
 * It is made freely available with the understanding that      *
 * every copy of this file must include this header and that    *
 * CSCS take no responsibility for the use of the enclosed      *
 * teaching material.                                           *
 *                                                              *
 * Purpose: Exchange ghost cell in 2 directions using a topology*
 *                                                              *
 * Contents: C-Source                                           *
 *                                                              *
 ****************************************************************/

/* Use only 16 processes for this exercise
 * Send the ghost cell in two directions: left<->right and top<->bottom
 * ranks are connected in a cyclic manner, for instance, rank 0 and 12 are connected
 *
 * process decomposition on 4*4 grid
 *
 * |-----------|
 * | 0| 1| 2| 3|
 * |-----------|
 * | 4| 5| 6| 7|
 * |-----------|
 * | 8| 9|10|11|
 * |-----------|
 * |12|13|14|15|
 * |-----------|
 *
 * Each process works on a 10*10 (SUBDOMAIN) block of data
 * the D corresponds to data, g corresponds to "ghost cells"
 * xggggggggggx
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * xggggggggggx
 */

#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

#define SUBDOMAIN 10
#define DOMAINSIZE (SUBDOMAIN+2)

int main(int argc, char *argv[])
{
    int rank, size, i, j, dims[2], periods[2], rank_top, rank_bottom, rank_left, rank_right;
    double data[DOMAINSIZE*DOMAINSIZE];
    MPI_Request request;
    MPI_Status status;
    MPI_Comm comm_cart;
    MPI_Datatype data_ghost;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    if (size!=16) {
        printf("please run this with 16 processors\n");
        MPI_Finalize();
        exit(1);
    }

    for (i=0; i<DOMAINSIZE*DOMAINSIZE; i++) {
        data[i]=rank;
    }

    // neighbouring ranks with cartesian grid communicator
    dims[0]=dims[1]=4;
    periods[0]=periods[1]=1;

    // we do not allow the reordering of ranks here
    // an alternative solution would be to allow the reordering and to use the new communicator for the communication
    // then the MPI library has the opportunity to choose the best rank order with respect to performance
    // CREATE a cartesian communicator (4*4) with periodic boundaries and use it to find your neighboring
    // ranks in all dimensions in a cyclic manner.
    MPI_Cart_create(MPI_COMM_WORLD, 2, dims, periods, 0, &comm_cart);
    MPI_Cart_shift(comm_cart, 0, 1, &rank_top, &rank_bottom);
    MPI_Cart_shift(comm_cart, 1, 1, &rank_left, &rank_right);

    //  derived datatype, create a datatype for sending the column
    MPI_Type_vector(SUBDOMAIN, 1, DOMAINSIZE, MPI_DOUBLE, &data_ghost);
    MPI_Type_commit(&data_ghost);

    //  ghost cell exchange with the neighbouring cells in all directions
    //  to the top
    MPI_Irecv(&data[2-1+(DOMAINSIZE-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_bottom, 0, MPI_COMM_WORLD, &request);
    MPI_Send(&data[2-1+(2-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_top, 0, MPI_COMM_WORLD);
    MPI_Wait(&request, &status);
    //  to the bottom
    MPI_Irecv(&data[2-1+(1-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_top, 0, MPI_COMM_WORLD, &request);
    MPI_Send(&data[2-1+(DOMAINSIZE-1-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_bottom, 0, MPI_COMM_WORLD);
    MPI_Wait(&request, &status);
    //  to the left
    MPI_Irecv(&data[DOMAINSIZE-1+(2-1)*DOMAINSIZE], 1, data_ghost, rank_right, 0, MPI_COMM_WORLD, &request);
    MPI_Send(&data[2-1+(2-1)*DOMAINSIZE], 1, data_ghost, rank_left, 0, MPI_COMM_WORLD);
    MPI_Wait(&request, &status);
    //  to the right
    MPI_Irecv(&data[1-1+(2-1)*DOMAINSIZE], 1, data_ghost, rank_left, 0, MPI_COMM_WORLD, &request);
    MPI_Send(&data[DOMAINSIZE-1-1+(2-1)*DOMAINSIZE], 1, data_ghost, rank_right, 0, MPI_COMM_WORLD);
    MPI_Wait(&request, &status);

    if (rank==9) {
        printf("data of rank 9 after communication\n");
        for (j=0; j<DOMAINSIZE; j++) {
            for (i=0; i<DOMAINSIZE; i++) {
                printf("%.1f ", data[i+j*DOMAINSIZE]);
            }
            printf("\n");
        }
    }

    MPI_Type_free(&data_ghost);
    MPI_Comm_free(&comm_cart);
    MPI_Finalize();

    return 0;
}
